### PR TITLE
feat(migrations): Final failure fixes for Distributed Testing Mode

### PR DESCRIPTION
**Pull request by ayirr7**
This PR contains the fixes and revisions for the remaining failure cases that were occurring in distributed testing mode (i.e. when `SNUBA_SETTINGS=test_distributed`). If these work as expected and are approved, then we can run this mode in CI.

### CODE REVIEW

File: snuba/migrations/snuba_migrations/outcomes/0001_outcomes.py [hunk 1]
```diff
@@ -125,4 +125,7 @@ def backwards_dist(self) -> Sequence[operations.SqlOperation]:
             operations.DropTable(
                 storage_set=StorageSetKey.OUTCOMES, table_name="outcomes_raw_dist",
             ),
+            operations.DropTable(
+                storage_set=StorageSetKey.OUTCOMES, table_name="outcomes_hourly_dist",
+            ),
```
Comment (lynnagara) [inline snuba/migrations/snuba_migrations/outcomes/0001_outcomes.py:130]: I think this needs to replace the drop table statement on lines 122-124, which incorrectly reference the _local table.

File: tests/datasets/test_sessions_processing.py [hunk 2]
```diff
@@ -98,7 +102,7 @@ def query_runner(
                 ["started", ">", "2020-01-01 12:00:00"],
             ],
         },
-        "sessions_hourly_local",
+        materialized_storage.get_schema().get_table_name(),
```
Comment (lynnagara) [inline tests/datasets/test_sessions_processing.py:105]: The read schema is already imported, why not simplify this to `read_schema.get_table_name()`

File: tests/migrations/test_operations.py [hunk 3]
```diff
@@ -37,7 +37,8 @@ def test_create_table() -> None:
                 settings={"index_granularity": "256"},
             ),
         ).format_sql()
-        == "CREATE TABLE IF NOT EXISTS test_table (id String, name Nullable(String), version UInt64) ENGINE ReplacingMergeTree(version) ORDER BY version SETTINGS index_granularity=256;"
+        == "CREATE TABLE IF NOT EXISTS test_table (id String, name Nullable(String), version UInt64) ENGINE ReplicatedReplacingMergeTree(version) ORDER BY version SETTINGS index_granularity=256;"
+        or "CREATE TABLE IF NOT EXISTS test_table (id String, name Nullable(String), version UInt64) ENGINE ReplicatedReplacingMergeTree('/clickhouse/tables/events/{shard}/snuba_test/test_table', '{replica}', version) ORDER BY version SETTINGS index_granularity=256;"
```
Comment (lynnagara) [inline tests/migrations/test_operations.py:41]: This isn't actually testing anything except the second condition on line 41 is not falsy, and won't fail even if the wrong SQL is produced

File: snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py [hunk 4]
```diff
@@ -198,5 +198,8 @@ def backwards_dist(self) -> Sequence[operations.SqlOperation]:
         return [
             operations.DropTable(
                 storage_set=StorageSetKey.EVENTS, table_name="errors_dist_new"
-            )
+            ),
+            operations.DropTable(
+                storage_set=StorageSetKey.EVENTS_RO, table_name="errors_dist_ro"
+            ),
```
Comment (lynnagara) [inline snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py:204]: Thanks

File: tests/migrations/test_runner_individual.py [hunk 5]
```diff
@@ -39,67 +39,71 @@ def teardown_function() -> None:
...
-    def get_sampling_key() -> Any:
-        database = cluster.get_database()
-        ((sampling_key,),) = perform_select_query(
-            ["sampling_key"],
-            "system.tables",
-            {"name": "transactions_local", "database": str(database)},
-            None,
-            connection,
+    if cluster.is_single_node():
+        connection = cluster.get_query_connection(ClickhouseClientSettings.MIGRATE)
```
Comment (lynnagara) [inline tests/migrations/test_runner_individual.py:44]: It's a bit more readable to just early return if it's not a single node and not increase the indentation of the whole test which is very long, i.e.
```
if cluster.is_single_node() is False:
    return
```
Please also add a comment about why we don't want to run the test in distributed mode.

File: tests/migrations/test_runner_individual.py [hunk 6]
```diff
@@ -134,45 +138,47 @@ def generate_transactions() -> None:
...
-        ORDER BY id SAMPLE BY id SETTINGS index_granularity = 8192
-        """
-    )
-
-    migration_id = "0010_groupedmessages_onpremise_compatibility"
 
-    runner = Runner()
-    runner.run_migration(MigrationKey(MigrationGroup.SYSTEM, "0001_migrations"))
-    events_migrations = get_group_loader(MigrationGroup.EVENTS).get_migrations()
+    if cluster.is_single_node():
```
Comment (lynnagara) [inline tests/migrations/test_runner_individual.py:142]: Similar to the other test, just early return if it's not a single node.

**Review by lynnagara — COMMENTED** [review]
Yay, almost there!

File: snuba/migrations/snuba_migrations/outcomes/0001_outcomes.py (hunk 1 above)
Comment (ayirr7) [reply snuba/migrations/snuba_migrations/outcomes/0001_outcomes.py:130]: Ah yes, good catch.

File: tests/datasets/test_sessions_processing.py (hunk 2 above)
Comment (ayirr7) [reply tests/datasets/test_sessions_processing.py:105]: Done.

File: tests/migrations/test_operations.py (hunk 3 above)
Comment (ayirr7) [reply tests/migrations/test_operations.py:41]: Oops...

File: tests/migrations/test_runner_individual.py (hunk 5 above)
Comment (ayirr7) [reply tests/migrations/test_runner_individual.py:44]: Ok!

File: tests/migrations/test_operations.py [hunk 7]
```diff
@@ -38,6 +38,19 @@ def test_create_table() -> None:
...
+            StorageSetKey.EVENTS,
+            "test_table",
+            columns,
+            ReplacingMergeTree(
+                storage_set=StorageSetKey.EVENTS,
+                version_column="version",
+                order_by="version",
+                settings={"index_granularity": "256"},
+            ),
+        ).format_sql()
```
Comment (lynnagara) [inline tests/migrations/test_operations.py:52]: Let's avoid duplicating all of this. You could rewrite this to:
```
assert CreateTable(...) in ["first_string", "second_string"]
```

Comment (ayirr7) [reply tests/migrations/test_operations.py:52]: Oh yes that'd be much cleaner

**Review by lynnagara — APPROVED** [review]

**Pull request merged**